TechnicalDocuments/Proposals/FlowAsStruct.md

```diff
@@ -0,0 +1,73 @@
+# Refactor `Flow` to be a Struct
+
+* Author: Martin Nygren
+* Review Manager: Anders Ha
+
+## Introduction
+
+This proposal is inspired by the [Pointfree](https://www.pointfree.co) episodes on Witness Oriented design, [33](https://www.pointfree.co/episodes/ep33-protocol-witnesses-part-1), [34](https://www.pointfree.co/episodes/ep34-protocol-witnesses-part-2), [35](https://www.pointfree.co/episodes/ep35-advanced-protocol-witnesses-part-1), [36](https://www.pointfree.co/episodes/ep36-advanced-protocol-witnesses-part-2) and [39](https://www.pointfree.co/episodes/ep39-witness-oriented-library-design).
+
+## Motivation
+
+Currently `Flow` is defined as a `Protocol` and implemented by a couple of helper classes (e.g. `EmptyFlow`, `NavigationFlow`, `ModalFlow`) that are made available as calculated properties (`Flow.empty`, `UINavigationController.navigationFlow`, `UIViewController.modalFlow`). This implementation already indicates that modelling the concept of a `Flow` as struct with function variables is closer to what we want then defining `Flow` as a protocol. We would have preferred to extend various `UIKit` classes if modelling `Flow` as a protocol is a natural choice.
```

Comment:
> We would have preferred to extend various `UIKit` classes if modelling `Flow` as a protocol is a natural choice.

Not necessarily. There are two key reasons:

1. They hide the nature of the presenter & may also hold a weak reference to it, so as not to exhibit retain cycles. This is especially important since the presenter behind `Flow` is pretty much guaranteed to have retained the view model observed by the flow controller.

    Arguably speaking if the `Flow` protocol is constrained to `AnyObject`, flow controllers could hold a weak reference to `Flow`. But then it makes custom `Flow` implementations impossible if we universally enforce this, as they would be deallocated right away unless being retained in other places. If we don't enforce this, we are breaking the interchangeability by binding the slots with a specific implementation detail.

2. `Flow` requirements have similar names to existing UIKit signatures. Direct conformance would clutter autocompletion, especially since it covers also more than UIViewController modal presentation which guarantees mismatches of method signatures.

Reply:
As we [have seen](https://www.pointfree.co/episodes/ep33-protocol-witnesses-part-1), protocols are not the only way to define abstractions.

Reply:
It is not the only way. I am just saying it is the blessed way by the language.